### DISCUSSION TITLE
Accessing indicator and model prediction data for plotting

### DISCUSSION THREAD

**Discussion by SW4T400** [edtechre/pybroker#77] [Q&A] [answered]
Hi @edtechre  @albertandking & pybroker community!

Question and Background
-------------------------------

I was wondering if there is a way to access the indicator & model prediction data / columns that the strategy class uses. Similar to the following providing a ```TestResult``` class, I would like to also access the underlying data that produced the results.
```
result = strategy.backtest()
```
I am asking because I would like to plot 

1. the ```DataSource``` alongside 
2. the indicators and 
3. the models prediction values 
4. as well as all the trades (this I get from the TestResult Class without problem)

My idea is to allow plotting for all 4 stages with only 3 and 4 requiring a backtest. For example I might want to plot only the basic data with the indicators, just to develop an indicator. Later in the development process I would like to run a backtest to develop trading rules and/or ML models. At this point I would like to not only see the trades but also the models prediction values that caused the trades.

Currently I only know how to achieve 1. & 2. - **can anyone tell me of a better way and/or tell me how to achieve 3. & 4.?**
```
indicator_set = IndicatorSet()
indicator_set.add(cmma_20, hhv_5)
df_indicators=indicator_set(df)

all_in_one_df=pd.merge(df, df_indicators, on=['date', 'symbol'])
```
This gives me a simple dataframe to plot. By filtering for a specific symbol I could plot its price and some indicators.

For 3. I assume I have to start from the following lines in the [tutorial](https://www.pybroker.com/en/latest/notebooks/6.%20Training%20a%20Model.html#Train-and-Backtest):

```
model_slr = pybroker.model('slr', train_slr, indicators=[cmma_20])

config = StrategyConfig(bootstrap_sample_size=100)
strategy = Strategy(YFinance(), '3/1/2017', '3/1/2022', config)
strategy.add_execution(None, ['NVDA', 'AMD'], models=model_slr)

strategy.backtest(train_size=0.5) 
```
The last line in the code above ↑ seems to perform the training and testing, but obviously does not give any trades (which I dont need for stage 3. anyways as I just want to look at prediction values alongside the price). Could I extract the prediction values (for all symbols) from this line (somehow) or am I on the wrong track?

**What would be the best way to get the ```DataSource``` columns, indicator & prediction values (plus the ```TradeResults```) all together ? - where do I start looking?**

Plotting
-------------------------------

I am wondering if any of you have already developed plotting solutions for pybroker?

### Here´s what I am trying to do..

I am using 1 minute bars and the only thing I found capable of plotting millions up to hundreds of millions of points  is [plotly-resampler](https://github.com/predict-idlab/plotly-resampler/tree/main). This works only for scatter plots, so I can show candlesticks only when the user has zoomed in enough for regular plotly to handle it. I am currenlty writing a dash app in order to display the data from 1. - 4. in my browser  (**but as mentioned I dont know how to access this data)**

My hope is that I will manage to get visualizations per symbol (showing price, indicators, prediction values & trades) as well as portfolio views.

I find the plotting of [backtesting.py](https://kernc.github.io/backtesting.py/#example) and [autotrader](https://github.com/kieran-mackle/AutoTrader) quite pleasing and am aiming for something similar.
- [indicator view](https://autotrader.readthedocs.io/en/latest/features/visualisation.html#interactive-visualisation-with-autotrader)
- [trading view](https://autotrader.readthedocs.io/en/latest/features/backtesting.html#single-bot-backtest)

What are your thoughts on this? Im new to both coding & algotrading and would appreciate any help and/or comments.

**Comment by edtechre** [comment, marked as answer]
Note, you can use the [return_signals](https://www.pybroker.com/en/latest/reference/pybroker.config.html#pybroker.config.StrategyConfig.return_signals) option to return bar data, indicator data, and prediction values.
